### ISSUE TITLE
[Feature] - 3D Terrain mode

### ISSUE THREAD

**Issue by jo-chemla** [Flomp/wanderer#111]
Hi there, 
First of all congrats on the amazing work here, first self-hosted app that gathers almost all features trekkers and bikepackers need, really impressive! 

It could be nice to think about adding 3D mode support, to easier see elevation gain for past/planned routes. Web or mobile apps that allow this feature: [Fatmap onw bought by Strava](https://about.fatmap.com/journal-digest/track-all-your-adventures-with-fatmap), [AllTrails+](https://support.alltrails.com/hc/fr/articles/360058630271-Comment-afficher-les-cartes-d-AllTrails-en-3D) [Komoot premium](https://www.komoot.com/fr-fr/premium/3d-maps) [OutdoorActive](https://www.outdooractive.com/fr/knowledgepage/apercu-flyover/37926720/)

### Formats
Elevation source has to be taken into consideration, when thinking about which library to include. Standard supported formats for elevation include: 
 - Maptiler [terrain-rgb](https://documentation.maptiler.com/hc/en-us/articles/4405444055313-RGB-Terrain-by-MapTiler) spec is a simple XYZ/TMS tiles file-server, where RGB encodes elevation via this conversion formula: `elevation = -10000 + ((R * 256 * 256 + G * 256 + B) * 0.1)`. 
 - Cesium [Quantized-mesh](https://github.com/CesiumGS/quantized-mesh) terrain format stores tiles at `https://server-endpoint/path/{z}/{x}/{y}.terrain`
 - Note Mapbox also has another terrain format [mapbox-terrain-v2](https://docs.mapbox.com/data/tilesets/reference/mapbox-terrain-v2/) which serves vector tiles rather than raster for elevation

### 2.5D Solutions
Leaflet is purely 2D, but here are a few libraries options - I don't know Svelte at all, so not sure which would best fit with the current stack: 
 - DeckGL/Loadersgl supports [terrain-layer](https://deck.gl/docs/api-reference/geo-layers/terrain-layer) following the `terrain-rgb` spec URL templates, see this [track+terrain example](https://deck.gl/examples/terrain-extension) as well as a simple [terrain example](https://deck.gl/examples/terrain-layer). Deckgl also supports [quantized-mesh terrain](https://loaders.gl/docs/modules/terrain/api-reference/quantized-mesh-loader). What's interesting with deckgl is that it supports a vast range of data sources, among which 3D-Tiles - so could load the [Google 3D Cities](https://deck.gl/examples/google-3d-tiles) dataset - [line-layer](https://deck.gl/examples/line-layer), [geojson layer path](https://deck.gl/examples/geojson-layer-paths), [heatmap](https://deck.gl/examples/heatmap-layer) etc
 - [MapLibre with 3D Terrain](https://www.maptiler.com/news/2022/05/maplibre-v2-add-3d-terrain-to-your-map/) (also [example](https://maplibre.org/maplibre-gl-js/docs/examples/3d-terrain/)) where terrainSource is a `terrain-rgb` source. Since maplibre forked mapbox following the licence change controversy, maplibre does not seem to support the mapbox-terrain-v2 spec. Multiple terrain-rgb template urls available like [this one](https://api.mapbox.com/v4/mapbox.terrain-rgb/{z}/{x}/{y}@2x.webp?sku=101FWQkypq8FN&access_token=pk.eyJ1Ijoic3ZjLW9rdGEtbWFwYm94LXN0YWZmLWFjY2VzcyIsImEiOiJjbG5sMnFlZ28wb2d5MnFtb2xnMG90OW96In0.IE8Vqs0NTzCY0WqPzV9kcw) from mapbox with a public access-token.
 - [CesiumJS with WorldTerrain](https://cesium.com/platform/cesium-ion/content/cesium-world-terrain/) stored as `quantized-mesh`, streamed from Cesium Ion - url template is [this one](https://assets.ion.cesium.com/us-east-1/asset_depot/1/CesiumWorldTerrain/v1.2/{z}/{x}/{y}.terrain?extensions=metadata&v=1.2.0). Also supports some other terrain providers like [ArcGISTiledElevation](https://cesium.com/learn/cesiumjs/ref-doc/ArcGISTiledElevationTerrainProvider.html)
 - [Giro3D](https://gitlab.com/giro3d/giro3d) by oslandia, is an extensive library made to support loads of geospatial & 3d formats, originally a fork of [iTowns](https://www.itowns-project.org/).

Note this [kylebarron/dem-tiler](https://github.com/kylebarron/dem-tiler) utility is useful to convert between terrain sources, since it can output both `terrain-rgb` or `quantized-mesh`, and reads [**open-data AWS Terrain tiles**](https://registry.opendata.aws/terrain-tiles/) or mosaic of COGs) in case you want to build a terrain dataset yourself rather than relying on some of the provided sources. 

## Full 3d support
The above lists 2.5D terrain formats. If you'd prefer full 3d-support, a tiled mesh format is preferred, like OGC (Open Geospatial Consortium) [3D-Tiles](https://cesium.com/why-cesium/3d-tiles/) inceptionned by cesium. Some data-vendrs offer data in these formats, like [Google 3D-Cities/Photorealistic 3D Tiles](https://mapsplatform.google.com/resources/blog/create-immersive-3d-map-experiences-photorealistic-3d-tiles/) available via api-key, free for all at-the-moment. 
Multiple of the above libs do support OGC 3dTiles, like DeckGL/loadersgl, CesiumJS, Giro3D etc. 

A few interesting frameworks based on some of the above-mentioned libs: [TerriaJS](https://www.terria.io/), [Piero/Giro3d](https://giro3d.org/piero.html)

**Comment by Flomp** [comment]
Thanks for the detailed issue and sorry that I haven't gotten back to you sooner. Looking into it briefly, this seems like a big undertaking which I currently lack the time to implement. 

In case you would like to try your luck: most maps in wanderer use [this component](https://github.com/Flomp/wanderer/blob/main/web/src/lib/components/trail/map_with_elevation.svelte).

I will leave the issue open, because it is a feature that I would ultimately really like to add.

**Comment by tofublock** [comment]
In the past, I have written a Django server that looks through locally available elevation data raster images, stitches them together as required, and returns a new raster image of the requested area. It was intended for a similar purpose (displaying terrain and overlay GPX tracks), but the project never went anywhere.
It's been a while, so I'd need to refresh my memory, but maybe this can be of help here. @Flomp, let me know if you think it might, and we can take a look.

**Comment by Flomp** [comment]
Hi there,

just coming back here to announce that this is implemented as of v0.12.0 (please be aware that there are breaking changes in this release. Checkout the [changelog](https://wanderer.to/getting-started/changelog/))

Here are the docs:
[https://wanderer.to/guides/customize-map/#terrain--hillshading](https://wanderer.to/guides/customize-map/#terrain--hillshading)
And here is a demo:
[https://demo.wanderer.to/map/trail/jmolew6qjtkmx5f](https://demo.wanderer.to/map/trail/jmolew6qjtkmx5f)

**Comment by jo-chemla** [comment]
Nice to see wanderer now supports 2.5D terrain mode via TileJSON following [Mapbox terrain v2](https://docs.mapbox.com/data/tilesets/reference/mapbox-terrain-v2/) spec, congrats on that! 
On my end, I never really used tilejson much for terrain since the only provider I've come across that do use it is mapbox, but it's great to see the feature available within wanderer!

Reading the changelog, I also tried loading a few different `style.json` but for the basemap rather than terrain+hillshade. I tried to load eg 
 - [Mapotempo/satellite-hybrid-gl-style](https://github.com/Mapotempo/satellite-hybrid-gl-style/blob/master/style.json),
 - also tried using Azavea https://tilejson.io but not sure whether the resulting styles follow the mapbox spec or another one. 
 
but no raster style seem to load. Have you managed to load `raster` tilesets via a style.json spec or does this work only for vector tilesets? Thanks as always for all your work!
